File: firmware/props/respawn_station_dual_team/test_harness.py
#!/usr/bin/env python3
"""
test_harness.py

Desktop test harness for the dual-team respawn station logic.
Run: python test_harness.py

This simulates both RED and BLUE units and validates:
- Ready state transitions
- Countdown logic
- Game start flow
- Count increments and limit checks
- Peer packet handling
"""

from enum import IntEnum
from dataclasses import dataclass
from typing import Optional

class Team(IntEnum):
    RED = 0
    BLUE = 1

class GameMode(IntEnum):
    RECORD_ONLY = 0
    KILL_LIMIT = 1
    RESPAWN_LIMIT = 2
    FLAG_CAPTURE = 3

class GameState(IntEnum):
    IDLE = 0
    READY = 1
    COUNTDOWN = 2
    RUNNING = 3
    GAMEOVER = 4

@dataclass
class Unit:
    device_id: str
    local_team: Team
    game_mode: GameMode = GameMode.RECORD_ONLY
    game_state: GameState = GameState.IDLE
    
    admin_enabled: bool = True
    red_ready: bool = False
    blue_ready: bool = False
    
    red_count: int = 0
    blue_count: int = 0
    
    limit_value: int = 20
    countdown_seconds: int = 10
    respawn_delay_seconds: int = 5
    
    last_message: str = "BOOT"
    game_over_reason: str = ""
    countdown_start_ms: int = 0
    
    def team_name(self) -> str:
        return "RED" if self.local_team == Team.RED else "BLUE"
    
    def game_state_name(self) -> str:
        names = {
            GameState.IDLE: "IDLE",
            GameState.READY: "READY",
            GameState.COUNTDOWN: "COUNTDOWN",
            GameState.RUNNING: "RUNNING",
            GameState.GAMEOVER: "GAME OVER",
        }
        return names.get(self.game_state, "UNKNOWN")
    
    def game_mode_name(self) -> str:
        names = {
            GameMode.RECORD_ONLY: "RECORD ONLY",
            GameMode.KILL_LIMIT: "KILL LIMIT",
            GameMode.RESPAWN_LIMIT: "RESPAWN LIMIT",
            GameMode.FLAG_CAPTURE: "FLAG CAPTURE",
        }
        return names.get(self.game_mode, "UNKNOWN")
    
    def mark_ready(self):
        if not self.admin_enabled or self.game_state == GameState.GAMEOVER:
            self.last_message = "READY_REJECTED"
            return
        
        if self.local_team == Team.RED:
            self.red_ready = True
        else:
            self.blue_ready = True
        
        self.game_state = GameState.READY
        self.last_message = f"{self.team_name()} READY"
    
    def handle_peer_ready(self, sender_team: Team):
        if sender_team == self.local_team:
            return
        
        if sender_team == Team.RED:
            self.red_ready = True
        else:
            self.blue_ready = True
        
        if self.red_ready and self.blue_ready:
            self.game_state = GameState.COUNTDOWN
            self.countdown_start_ms = 0
            self.last_message = "COUNTDOWN_START"
    
    def tick_countdown(self, elapsed_ms: int):
        if self.game_state != GameState.COUNTDOWN:
            return
        
        elapsed_seconds = elapsed_ms // 1000
        if elapsed_seconds >= self.countdown_seconds:
            self.game_state = GameState.RUNNING
            self.last_message = "GAME START"
    
    def register_action_press(self):
        if not self.admin_enabled or self.game_state != GameState.RUNNING:
            self.last_message = "ACTION_REJECTED"
            return
        
        if self.local_team == Team.RED:
            self.red_count += 1
            count = self.red_count
        else:
            self.blue_count += 1
            count = self.blue_count
        
        self.last_message = f"COUNT_{count}"
        
        # Check limits
        if self.game_mode == GameMode.KILL_LIMIT:
            if self.red_count >= self.limit_value:
                self.game_state = GameState.GAMEOVER
                self.game_over_reason = "RED HIT KILL LIMIT"
                self.last_message = "GAMEOVER"
            if self.blue_count >= self.limit_value:
                self.game_state = GameState.GAMEOVER
                self.game_over_reason = "BLUE HIT KILL LIMIT"
                self.last_message = "GAMEOVER"
    
    def handle_peer_count(self, value: str):
        if value.startswith("RED:"):
            self.red_count = int(value[4:])
        elif value.startswith("BLUE:"):
            self.blue_count = int(value[5:])
    
    def reset(self):
        self.red_ready = False
        self.blue_ready = False
        self.red_count = 0
        self.blue_count = 0
        self.game_over_reason = ""
        self.game_state = GameState.IDLE
        self.last_message = "RESET"
    
    def status(self):
        print(f"  [{self.device_id}] Team: {self.team_name()} | State: {self.game_state_name()} | Mode: {self.game_mode_name()} | Counts: R={self.red_count} B={self.blue_count} | Last: {self.last_message}")


def test_ready_state_transition():
    print("TEST 1: Ready State Transition")
    red = Unit("RESPAWN-RED-01", Team.RED)
    blue = Unit("RESPAWN-BLUE-01", Team.BLUE)
    
    red.status()
    blue.status()
    assert red.game_state == GameState.IDLE
    assert blue.game_state == GameState.IDLE
    
    red.mark_ready()
    print("  RED marked ready")
    red.status()
    assert red.game_state == GameState.READY
    assert red.red_ready == True
    
    blue.mark_ready()
    print("  BLUE marked ready")
    blue.status()
    assert blue.game_state == GameState.READY
    
    # Peer sync
    red.handle_peer_ready(Team.BLUE)
    blue.handle_peer_ready(Team.RED)
    print("  Peer sync complete")
    red.status()
    blue.status()
    assert red.game_state == GameState.COUNTDOWN
    assert blue.game_state == GameState.COUNTDOWN
    print("  ✓ PASS\n")


def test_countdown_and_game_start():
    print("TEST 2: Countdown and Game Start")
    red = Unit("RESPAWN-RED-01", Team.RED)
    blue = Unit("RESPAWN-BLUE-01", Team.BLUE)
    
    red.countdown_seconds = 3
    blue.countdown_seconds = 3
    red.game_state = GameState.COUNTDOWN
    blue.game_state = GameState.COUNTDOWN
    
    red.tick_countdown(0)
    blue.tick_countdown(0)
    assert red.game_state == GameState.COUNTDOWN
    
    red.tick_countdown(3500)  # Tick past 3 seconds
    blue.tick_countdown(3500)
    print("  After countdown expiry:")
    red.status()
    blue.status()
    assert red.game_state == GameState.RUNNING
    assert blue.game_state == GameState.RUNNING
    print("  ✓ PASS\n")


def test_action_presses_and_count_sync():
    print("TEST 3: Action Presses and Count Sync")
    red = Unit("RESPAWN-RED-01", Team.RED)
    blue = Unit("RESPAWN-BLUE-01", Team.BLUE)
    
    red.game_mode = GameMode.KILL_LIMIT
    blue.game_mode = GameMode.KILL_LIMIT
    red.limit_value = 5
    blue.limit_value = 5
    red.game_state = GameState.RUNNING
    blue.game_state = GameState.RUNNING
    
    for i in range(3):
        red.register_action_press()
        blue.handle_peer_count(f"RED:{red.red_count}")
    
    print("  RED pressed 3 times:")
    red.status()
    blue.status()
    assert red.red_count == 3
    assert blue.red_count == 3
    
    for i in range(5):
        blue.register_action_press()
        red.handle_peer_count(f"BLUE:{blue.blue_count}")
    
    print("  BLUE pressed 5 times (hit limit):")
    red.status()
    blue.status()
    assert blue.game_state == GameState.GAMEOVER
    assert "BLUE HIT KILL LIMIT" in blue.game_over_reason
    assert red.blue_count == 5
    print("  ✓ PASS\n")


def test_reset_and_restart():
    print("TEST 4: Reset and Restart")
    red = Unit("RESPAWN-RED-01", Team.RED)
    blue = Unit("RESPAWN-BLUE-01", Team.BLUE)
    
    red.game_state = GameState.GAMEOVER
    blue.game_state = GameState.GAMEOVER
    red.red_count = 10
    blue.blue_count = 10
    
    red.reset()
    blue.reset()
    print("  After reset:")
    red.status()
    blue.status()
    assert red.game_state == GameState.IDLE
    assert blue.game_state == GameState.IDLE
    assert red.red_count == 0
    assert blue.blue_count == 0
    print("  ✓ PASS\n")


def test_admin_disable():
    print("TEST 5: Admin Disable")
    red = Unit("RESPAWN-RED-01", Team.RED)
    
    red.admin_enabled = False
    red.mark_ready()
    print("  Attempted ready while disabled:")
    red.status()
    assert red.game_state == GameState.IDLE
    assert red.last_message == "READY_REJECTED"
    print("  ✓ PASS\n")


def test_record_only_mode():
    print("TEST 6: Record-Only Mode (No Limit)")
    red = Unit("RESPAWN-RED-01", Team.RED)
    blue = Unit("RESPAWN-BLUE-01", Team.BLUE)
    
    red.game_mode = GameMode.RECORD_ONLY
    blue.game_mode = GameMode.RECORD_ONLY
    red.limit_value = 5
    blue.limit_value = 5
    red.game_state = GameState.RUNNING
    blue.game_state = GameState.RUNNING
    
    for i in range(10):
        red.register_action_press()
        blue.handle_peer_count(f"RED:{red.red_count}")
    
    print("  RED pressed 10 times (exceeds limit but record-only):")
    red.status()
    blue.status()
    assert red.game_state == GameState.RUNNING  # Still running
    assert red.red_count == 10
    print("  ✓ PASS\n")


if __name__ == "__main__":
    print("=== AOJ Dual-Team Respawn Station Test Harness ===\n")
    
    try:
        test_ready_state_transition()
        test_countdown_and_game_start()
        test_action_presses_and_count_sync()
        test_reset_and_restart()
        test_admin_disable()
        test_record_only_mode()
        
        print("=== All Tests Passed ===")
    except AssertionError as e:
        print(f"❌ TEST FAILED: {e}")
        exit(1)
    except Exception as e:
        print(f"❌ ERROR: {e}")
        exit(1)

File: firmware/props/respawn_station_dual_team/test_harness.cpp
/**
 * test_harness.cpp
 * 
 * Desktop test harness for the dual-team respawn station logic.
 * Compile & run: g++ -std=c++17 -o test_harness test_harness.cpp && ./test_harness
 * 
 * This simulates both RED and BLUE units and validates:
 * - Ready state transitions
 * - Countdown logic
 * - Game start flow
 * - Count increments and limit checks
 * - Peer packet handling
 * - Server command routing
 */

#include <iostream>
#include <string>
#include <map>
#include <ctime>
#include <cassert>

using namespace std;

enum Team { TEAM_RED = 0, TEAM_BLUE = 1 };
enum GameMode { MODE_RECORD_ONLY = 0, MODE_KILL_LIMIT = 1, MODE_RESPAWN_LIMIT = 2, MODE_FLAG_CAPTURE = 3 };
enum GameState { STATE_IDLE = 0, STATE_READY = 1, STATE_COUNTDOWN = 2, STATE_RUNNING = 3, STATE_GAMEOVER = 4 };

struct Unit {
  string deviceId;
  Team localTeam;
  GameMode gameMode;
  GameState gameState;
  
  bool adminEnabled;
  bool redReady;
  bool blueReady;
  
  int redCount;
  int blueCount;
  
  int limitValue;
  int countdownSeconds;
  int respawnDelaySeconds;
  
  string lastMessage;
  string gameOverReason;
  
  unsigned long countdownStartMs;
  
  Unit(string id, Team team) 
    : deviceId(id), localTeam(team), gameMode(MODE_RECORD_ONLY), gameState(STATE_IDLE),
      adminEnabled(true), redReady(false), blueReady(false),
      redCount(0), blueCount(0),
      limitValue(20), countdownSeconds(10), respawnDelaySeconds(5),
      lastMessage("BOOT"), countdownStartMs(0) {}
  
  string teamName() const { return localTeam == TEAM_RED ? "RED" : "BLUE"; }
  string gameStateName() const {
    switch (gameState) {
      case STATE_IDLE: return "IDLE";
      case STATE_READY: return "READY";
      case STATE_COUNTDOWN: return "COUNTDOWN";
      case STATE_RUNNING: return "RUNNING";
      case STATE_GAMEOVER: return "GAME OVER";
    }
    return "UNKNOWN";
  }
  
  string gameModeName() const {
    switch (gameMode) {
      case MODE_RECORD_ONLY: return "RECORD ONLY";
      case MODE_KILL_LIMIT: return "KILL LIMIT";
      case MODE_RESPAWN_LIMIT: return "RESPAWN LIMIT";
      case MODE_FLAG_CAPTURE: return "FLAG CAPTURE";
    }
    return "UNKNOWN";
  }
  
  void markReady() {
    if (!adminEnabled || gameState == STATE_GAMEOVER) {
      lastMessage = "READY_REJECTED";
      return;
    }
    if (localTeam == TEAM_RED) redReady = true;
    if (localTeam == TEAM_BLUE) blueReady = true;
    gameState = STATE_READY;
    lastMessage = teamName() + " READY";
  }
  
  void handlePeerReady(Team senderTeam) {
    if (senderTeam == localTeam) return;
    if (senderTeam == TEAM_RED) redReady = true;
    if (senderTeam == TEAM_BLUE) blueReady = true;
    
    if (redReady && blueReady) {
      gameState = STATE_COUNTDOWN;
      countdownStartMs = 0; // Simulated time
      lastMessage = "COUNTDOWN_START";
    }
  }
  
  void startCountdown() {
    if (redReady && blueReady && gameState == STATE_READY) {
      gameState = STATE_COUNTDOWN;
      countdownStartMs = 0;
      lastMessage = "COUNTDOWN";
    }
  }
  
  void tickCountdown(unsigned int elapsedMs) {
    if (gameState != STATE_COUNTDOWN) return;
    unsigned long elapsedSeconds = elapsedMs / 1000;
    if (elapsedSeconds >= (unsigned long)countdownSeconds) {
      gameState = STATE_RUNNING;
      lastMessage = "GAME START";
    }
  }
  
  void registerActionPress() {
    if (!adminEnabled || gameState != STATE_RUNNING) {
      lastMessage = "ACTION_REJECTED";
      return;
    }
    
    if (localTeam == TEAM_RED) {
      redCount++;
    } else {
      blueCount++;
    }
    lastMessage = "COUNT_" + to_string(localTeam == TEAM_RED ? redCount : blueCount);
    
    // Check limits
    if (gameMode == MODE_KILL_LIMIT) {
      if (redCount >= limitValue) {
        gameState = STATE_GAMEOVER;
        gameOverReason = "RED HIT KILL LIMIT";
        lastMessage = "GAMEOVER";
      }
      if (blueCount >= limitValue) {
        gameState = STATE_GAMEOVER;
        gameOverReason = "BLUE HIT KILL LIMIT";
        lastMessage = "GAMEOVER";
      }
    }
  }
  
  void handlePeerCount(string value) {
    if (value.find("RED:") == 0) {
      redCount = stoi(value.substr(4));
    } else if (value.find("BLUE:") == 0) {
      blueCount = stoi(value.substr(5));
    }
  }
  
  void reset() {
    redReady = false;
    blueReady = false;
    redCount = 0;
    blueCount = 0;
    gameOverReason = "";
    gameState = STATE_IDLE;
    lastMessage = "RESET";
  }
  
  void status() const {
    cout << "  [" << deviceId << "] Team: " << teamName() 
         << " | State: " << gameStateName() 
         << " | Mode: " << gameModeName()
         << " | Counts: R=" << redCount << " B=" << blueCount
         << " | Last: " << lastMessage << endl;
  }
};

int main() {
  cout << "=== AOJ Dual-Team Respawn Station Test Harness ===" << endl << endl;
  
  Unit red("RESPAWN-RED-01", TEAM_RED);
  Unit blue("RESPAWN-BLUE-01", TEAM_BLUE);
  
  // TEST 1: Both units become ready
  cout << "TEST 1: Ready State Transition" << endl;
  red.status();
  blue.status();
  assert(red.gameState == STATE_IDLE);
  assert(blue.gameState == STATE_IDLE);
  
  red.markReady();
  cout << "  RED marked ready" << endl;
  red.status();
  assert(red.gameState == STATE_READY);
  assert(red.redReady == true);
  
  blue.markReady();
  cout << "  BLUE marked ready" << endl;
  blue.status();
  assert(blue.gameState == STATE_READY);
  
  // Peer sync: each sees the other ready
  red.handlePeerReady(TEAM_BLUE);
  blue.handlePeerReady(TEAM_RED);
  cout << "  Peer sync complete" << endl;
  red.status();
  blue.status();
  assert(red.gameState == STATE_COUNTDOWN);
  assert(blue.gameState == STATE_COUNTDOWN);
  cout << "  ✓ PASS" << endl << endl;
  
  // TEST 2: Countdown and game start
  cout << "TEST 2: Countdown and Game Start" << endl;
  red.countdownSeconds = 3;
  blue.countdownSeconds = 3;
  
  red.tickCountdown(0);
  blue.tickCountdown(0);
  assert(red.gameState == STATE_COUNTDOWN);
  
  red.tickCountdown(3500); // Tick past 3 seconds
  blue.tickCountdown(3500);
  cout << "  After countdown expiry:" << endl;
  red.status();
  blue.status();
  assert(red.gameState == STATE_RUNNING);
  assert(blue.gameState == STATE_RUNNING);
  cout << "  ✓ PASS" << endl << endl;
  
  // TEST 3: Action presses and count sync
  cout << "TEST 3: Action Presses and Count Sync" << endl;
  red.gameMode = MODE_KILL_LIMIT;
  blue.gameMode = MODE_KILL_LIMIT;
  red.limitValue = 5;
  blue.limitValue = 5;
  
  for (int i = 0; i < 3; i++) {
    red.registerActionPress();
    blue.handlePeerCount("RED:" + to_string(red.redCount));
  }
  cout << "  RED pressed 3 times:" << endl;
  red.status();
  blue.status();
  assert(red.redCount == 3);
  assert(blue.redCount == 3);
  
  for (int i = 0; i < 5; i++) {
    blue.registerActionPress();
    red.handlePeerCount("BLUE:" + to_string(blue.blueCount));
  }
  cout << "  BLUE pressed 5 times (hit limit):" << endl;
  red.status();
  blue.status();
  assert(blue.gameState == STATE_GAMEOVER);
  assert(blue.gameOverReason.find("BLUE HIT KILL LIMIT") != string::npos);
  assert(red.blueCount == 5);
  cout << "  ✓ PASS" << endl << endl;
  
  // TEST 4: Reset and restart
  cout << "TEST 4: Reset and Restart" << endl;
  red.reset();
  blue.reset();
  cout << "  After reset:" << endl;
  red.status();
  blue.status();
  assert(red.gameState == STATE_IDLE);
  assert(blue.gameState == STATE_IDLE);
  assert(red.redCount == 0);
  assert(blue.blueCount == 0);
  cout << "  ✓ PASS" << endl << endl;
  
  // TEST 5: Admin disable
  cout << "TEST 5: Admin Disable" << endl;
  red.adminEnabled = false;
  blue.adminEnabled = false;
  red.markReady();
  cout << "  Attempted ready while disabled:" << endl;
  red.status();
  assert(red.gameState == STATE_IDLE);
  assert(red.lastMessage == "READY_REJECTED");
  cout << "  ✓ PASS" << endl << endl;
  
  // TEST 6: Record-only mode (no limit)
  cout << "TEST 6: Record-Only Mode (No Limit)" << endl;
  red.reset();
  blue.reset();
  red.adminEnabled = true;
  blue.adminEnabled = true;
  red.gameMode = MODE_RECORD_ONLY;
  blue.gameMode = MODE_RECORD_ONLY;
  red.limitValue = 5;
  blue.limitValue = 5;
  
  red.gameState = STATE_RUNNING;
  blue.gameState = STATE_RUNNING;
  
  for (int i = 0; i < 10; i++) {
    red.registerActionPress();
    blue.handlePeerCount("RED:" + to_string(red.redCount));
  }
  cout << "  RED pressed 10 times (exceeds limit but record-only):" << endl;
  red.status();
  blue.status();
  assert(red.gameState == STATE_RUNNING); // Still running
  assert(red.redCount == 10);
  cout << "  ✓ PASS" << endl << endl;
  
  cout << "=== All Tests Passed ===" << endl;
  return 0;
}
